utf8.h

```diff
@@ -87,187 +87,209 @@ extern "C" {
 #define utf8_null 0
 #endif
 
+#if defined(__cplusplus) && __cplusplus >= 201402L
+#define utf8_constexpr14 constexpr
+#else
+#define utf8_constexpr14
+#endif
+
+#if defined(__cplusplus) && __cplusplus >= 202002L
+using utf8_t = char8_t; // Introduced in C++20
```

Comment:
I think you should just introduce a `utf8_int8_t` here (similar to `utf8_int32_t`) and just use that everywhere. Like even for C do `typedef char utf8_int8_t`. That way you can even remove the internal type `utf8_t_` too?

I originally made all the entry points `void*` just because I didn't want people to think of the data as if it was a regular ASCII C String, but honestly with this PR you've very nicely solved the problem by introducing a new typedef!

Reply:
This will solve the problem indeed. However, I think it makes sense to keep `char8_t` for C++20. This type is specifically designed to refer to UTF8 data
If we do that, we can indeed remove the internal type